### ISSUE TITLE
panic on reduce + build-string

### ISSUE THREAD

**Issue by soruh** [nushell/nushell#3391]
**Describe the bug**
Running the following panics `nu`:
```nu
echo ["foo", "bar"] | reduce { = build-string $acc $it }
```
leads to
```
thread 'main' panicked at 'internal error: entered unreachable code', /home/soruh/.cargo/registry/src/github.com-1ecc6299db9ec823/nu-engine-0.30.0/src/evaluate/evaluator.rs:62:22
stack backtrace:
   0: rust_begin_unwind
             at /rustc/42816d61ead7e46d462df997958ccfd514f8c21c/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/42816d61ead7e46d462df997958ccfd514f8c21c/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at /rustc/42816d61ead7e46d462df997958ccfd514f8c21c/library/core/src/panicking.rs:50:5
   3: nu_engine::evaluate::evaluator::evaluate_baseline_expr
   4: nu_engine::evaluate::expr::run_expression_block
   5: nu_engine::evaluate::block::run_block
   6: nu_command::commands::reduce::process_row
   7: <nu_command::commands::reduce::Reduce as nu_engine::whole_stream_command::WholeStreamCommand>::run_with_actions
   8: nu_engine::whole_stream_command::WholeStreamCommand::run
   9: nu_engine::whole_stream_command::Command::run
  10: nu_engine::evaluation_context::EvaluationContext::run_command
  11: nu_engine::evaluate::block::run_block
  12: nu_engine::script::process_script
  13: nu_engine::script::run_script_standalone
  14: nu_cli::cli::run_script_file
  15: nu::main
```

**To Reproduce**
* run `nu -c 'echo ["foo", "bar"] | reduce { = build-string $acc $it }'`

**Expected behavior**
`nu` should not panic but return an error/"foobar" (I am guessing the command is wrong).

**Configuration**
```nu
> version | pivot
───┬────────────────────┬───────────────────────────────────────────────────────────────────────
 # │      Column0       │                                Column1
───┼────────────────────┼───────────────────────────────────────────────────────────────────────
 0 │ version            │ 0.30.0
 1 │ build_os           │ linux-x86_64
 2 │ rust_version       │ rustc 1.51.0 (2fd73fabe 2021-03-23)
 3 │ rust_channel       │ nightly-x86_64-unknown-linux-gnu (default)
 4 │ cargo_version      │ cargo 1.51.0 (43b129a20 2021-03-16)
 5 │ pkg_version        │ 0.30.0
 6 │ build_time         │ 2021-05-06 19:01:36
 7 │ build_rust_channel │ release
 8 │ features           │ ctrlc, default, directories, dirs, ichwh, rustyline, term, which, zip
───┴────────────────────┴───────────────────────────────────────────────────────────────────────
```
I have rebuilt `nu` with `cargo +stable` so `rust_channel` is inaccurate and should be `stable-x86_64-unknown-linux-gnu`

**Comment by soruh** [comment]
Looking at it in more detail it seems this is due to the same underlying issue as #3350 I'm happy to close this / move it to a comment if this doesn't add any additional information.

**Comment by jonathandturner** [comment]
Just tried this out and this is fixed by #3389

@soruh that's okay, they're different test cases so I don't mind keeping both issues open.

**Comment by jonathandturner** [comment]
For folks who come to this later, the current version works:

```
> ["foo", "bar"] | reduce { build-string $acc $it }
foobar
```
